org.knime.scijava.commands/src/org/knime/scijava/commands/module/DefaultNodeModule.java

```diff
@@ -115,71 +114,92 @@ private void preProcess(final DataRow input) throws Exception {
 
-        if (outputListener != null) {
-            outputListener.setCellOutput(output);
-            outputListener.setExecutionContext(ctx);
-        }
+        outputListener.setCellOutput(output);
+        outputListener.setExecutionContext(ctx);
 
         module.run();
 
-        if (outputListener == null) {
-            postProcess(output, ctx);
-        }
+        outputListener.flush();
     }
 
-    // internal usage only
-    class NodeModuleOutputChangedListener implements MultiOutputListener {
-
+    private class NodeModuleOutputChangedListener
+            implements MultiOutputListener {
         private ExecutionContext ctx;
 
         private CellOutput output;
 
-        public NodeModuleOutputChangedListener() {
-        }
+        /* true if the modules handles pushes itself, false otherwise */
+        private boolean manualPush = false;
 
         @Override
         public void notifyListener() {
             try {
-                postProcess(output, ctx);
+                // output can be null if sink node
+                if (output != null) {
+                    final List<DataCell> cells = new ArrayList<DataCell>();
+
+                    final ModuleItem<?> result = module.getInfo()
+                            .getOutput("result");
+                    if (result != null) {
+                        // FIXME hack because e.g. python script contains
+                        // result log
+                        cells.add(cs.convertToKnime(
+                                module.getOutput(result.getName()),
+                                result.getType(), outputMapping.get(result),
+                                ctx));
+                    }
+                    output.push(cells.toArray(new DataCell[cells.size()]));
+                }
             } catch (final Exception e) {
                 // FIXME
                 e.printStackTrace();
             }
         }
 
+        /**
+         * Set KNIME Execution Context
+         *
+         * @param ctx
+         *            execution context
+         */
         public void setExecutionContext(final ExecutionContext ctx) {
             this.ctx = ctx;
         }
 
+        /**
+         * Set CellOutput to push rows to.
+         *
+         * @param output
+         *            cell output
+         */
         public void setCellOutput(final CellOutput output) {
             this.output = output;
         }
+
+        /**
+         * Final flush. Will push a row in case module does not handle pushing
+         * output rows.
+         */
+        public void flush() {
+            if (!manualPush) {
+                notifyListener();
+            }
+        }
+
+        /**
+         * Enable/Disable manual push.
+         *
+         * @param b
+         *            if <code>true</code>, signals that the module handles
+         *            pushing rows.
+         */
+        public void enableManualPush(final boolean b) {
```

Comment:
Well, since it is only called immediatly after the constructor, I can put it into the constructor, makes sense :)